### PR TITLE
Ingela/ssl/tls record handling/gh 5961/otp 18087

### PR DESCRIPTION
**Pull request by IngelaAndin**
I think this should solve #5961  The problem seems to be that an optimization bug happened to not show itself until some other unrelated things happened to expose it. This has to do with how handshake messages are packed into TLS records. I will try to write a withe-box test to test it. I would also appreciate @michaelklishin, and @lukebakken  if you could also try it out.

### CODE REVIEW

**Comment by lukebakken** [comment]
I will pass it on to the user who reported this issue!

File: lib/ssl/src/tls_gen_connection.erl [hunk 1]
```diff
@@ -756,21 +754,27 @@ next_record_done(#state{protocol_buffers = Buffers} = State, CipherTexts, Connec
...
 %% (before the versions is actually negotiated) will always be the value of TLS-1.2 (which is a legacy
 %% field in TLS-1.3 client hello). The versions are instead negotiated with an hello extension. When
 %% decoding the server_hello messages we want to go through TLS-1.3 decode functions to be able
 %% to handle TLS-1.3 extensions if TLS-1.3 will be the negotiated version.
-effective_version({3,3} , #{versions := [{3,4} = Version |_]}, client, StateName) when StateName == hello;
-                                                                                       StateName == wait_sh ->
-    Version;
+handle_unnegotiated_version({3,3} , #{versions := [{3,4} = Version |_]} = Options, Data, Buffer, client, hello) ->
+    %% The effective version for decoding the server hello message should be the TLS-1.3. Possible coalesced TLS-1.2
+    %% server handshake messages should be decode with the negotiated version in later state.
```
Comment (JeromeDeBretagne) [inline lib/ssl/src/tls_gen_connection.erl:764]: Nitpick: "should be decoded"

Comment (IngelaAndin) [reply lib/ssl/src/tls_gen_connection.erl]: Thanks for the nitpick, fixed it

**Comment by IngelaAndin** [comment]
Turns out it is not that easy to write a test case without having a TLS-1.2 implementation that coalesces the handshake messages into one TLS-record.  I believe that change is correct and will work and it does not break any test cases. So I will merge this for OTP-25 and plan for test enhancement during OTP-26 development. If you can confirm the patch I will patch it on the OTP-24 track.

**Comment by IngelaAndin** [comment]
Ok, I will clear the patch for OTP-24. On the subject of LDAP, I am afraid I do not have much help to offer.

**Comment by lukebakken** [comment]
Hi @IngelaAndin. Well, it turns out that the issue is not resolved. See my comment here:

https://github.com/rabbitmq/rabbitmq-server/discussions/4726#discussioncomment-2784005

I can see that my LDAP test application works using 23.3.3 and I can see the same handshake issue reported using 24.3.4. When I use the `master` branch of `erlang/otp`, the crash doesn't happen but the test application never proceeds to doing the simple bind. It's as though the handshake does not complete.

Here is a TCP capture for port 636 for all three of the sessions in [the linked comment](https://github.com/rabbitmq/rabbitmq-server/discussions/4726#discussioncomment-2784005):

[ldap-port-636-tls-sessions.zip](https://github.com/erlang/otp/files/8730160/ldap-port-636-tls-sessions.zip)

The client and server certs can be found here:

https://github.com/lukebakken/rabbitmq-server-4726/tree/lukebakken/bakkenl-z01/certs

The certs being used have `bakkenl-z01` in the file name. Let me know if I can assist further.

**Pull request merged**